apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/RetryTests.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License").
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at 
+//
+// https://www.apache.org/licenses/LICENSE-2.0 
+//
+// Unless required by applicable law or agreed to in writing, software 
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using Google.Apis.Requests;
+using Google.Apis.Storage.v1;
+using System;
+using System.Net;
+using Xunit;
+
+namespace Google.Cloud.Storage.V1.Tests;
+
+/// <summary>
+/// Tests the complete retry flow with custom retry options provided
+/// </summary>
+public class RetryTests
+{
+    [Theory]
+    [InlineData(1, 6, 2, 3, 502, 502, 504)]
+    [InlineData(2, 6, 2, 6, 502, 502, 504)]
+    [InlineData(2, 6, 2, 0, 429, 502, 504)]
+    public void RetryTestsWithCustomValues(int initialBackoff, int maxBackoff, int backoffMultiplier, int totalTime, int statusCode, params int[] errorCodes)
+    {
+        RetryOptions retryOptions = new RetryOptions(
+            retryTiming: new RetryTiming(initialBackoff: TimeSpan.FromSeconds(initialBackoff),
+            maxBackoff: TimeSpan.FromSeconds(maxBackoff), backoffMultiplier: backoffMultiplier),
+            retryPredicate: RetryPredicate.FromErrorCodes(errorCodes));
+
+        Retry(service => service.Buckets.Get("bucket"), client => client.GetBucket("bucket", new GetBucketOptions { RetryOptions = retryOptions }), retryOptions, totalTime, (HttpStatusCode)statusCode);
+    }
+
+    private static void Retry<T>(Func<StorageService, ClientServiceRequest<T>> requestProvider, Action<StorageClient> clientAction,
```

Comment:
I would expect this to have:

- An array of expected times of requests (first of which would be zero)
- An array of status codes to return
- Nothing optional
- Don't use retryOptions other than to pass to GetBucketOptions
- Don't have the Func or Action - just hard code
- Either another parameter for "should expect eventual failure or success" or just check whether the final response code is 200